### PR TITLE
Enable double/triple buffered EDM channels for reduce-scatter 

### PR DESCRIPTION
**Pull request by caixunshiren**
### Ticket
#11853

### Problem description
Reduce scatter is currently not taking advantage of the multi-buffered channels feature of EDM. By taking advantage of this feature, reduce scatter should be reaching performance levels comparable to all-gather (for similar data sizes sent over the ethernet link).

This PR the equivalent change, for all-gather: https://github.com/tenstorrent/tt-metal/pull/11596

### What's changed
- Update the reduce scatter worker kernels to use the  ccl::edm::WorkerToEdmReader and ccl::edm::WorkerToEdmSender for the receiver and sender kernels, respectively.
- WIP

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp [hunk 1]
```diff
@@ -406,19 +408,73 @@ static void add_worker_config_to_edm_builders(
                     // Since we are in worker signal EDM termination mode, we don't need to set the actual number of
                     // messages the EDM must forward as it will receive its finish signal from the worker instead
                     1,
-                    receiver_worker_coords,
-                    expected_message_size_bytes);
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:410]: see comment above

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp [hunk 2]
```diff
@@ -386,12 +388,12 @@ static void add_worker_config_to_edm_builders(
...
                     worker_sender_semaphore_id,
                     1,  // cw_edm_channel_num_messages_to_send_per_transfer.at(c) * (ring_size - 1),
-                    sender_worker_coords,
-                    expected_message_size_bytes);
+                    sender_worker_coords);
             edm_interface_addresses.worker_sender_edm_semaphore_addresses.insert(
                 {global_worker_idx, sender_channel_buffer_info.eth_semaphore_l1_address});
             edm_interface_addresses.worker_sender_edm_buffer_addresses.insert(
                 {global_worker_idx, sender_channel_buffer_info.eth_buffer_l1_address});
+            sender_edm_builder.set_max_message_size_bytes(sender_channel_buffer_info.channel, clockwise_edm_builders.at(0).get_eth_buffer_size_bytes());
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:396]: should be `.at(link)`?

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp [hunk 3]
```diff
@@ -406,19 +408,73 @@ static void add_worker_config_to_edm_builders(
...
                 {global_worker_idx, receiver_channel_buffer_info.eth_semaphore_l1_address});
             edm_interface_addresses.worker_receiver_edm_buffer_addresses.insert(
                 {global_worker_idx, receiver_channel_buffer_info.eth_buffer_l1_address});
+            receiver_edm_builder.set_max_message_size_bytes(receiver_channel_buffer_info.channel, clockwise_edm_builders.at(0).get_eth_buffer_size_bytes());
         }
     }
 }
 
-static std::tuple<KernelHandle, KernelHandle> build_reduce_scatter_worker(
+static std::tuple<KernelHandle, KernelHandle, KernelHandle> build_reduce_scatter_worker_ct(
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:421]: Why is this change here? Is this the kernel creation optimization but this branch is based off that?

**Review by SeanNijjar — COMMENTED** [review]
I think I need to look at another PR first

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp (hunk 2 above)
Comment (caixunshiren) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:396]: I think all of them have the same eth_buffer_size?

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp (hunk 3 above)
Comment (caixunshiren) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:421]: Yes. This pr is based on the rt/ct arg optimization pr which has not been merged yet

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp (hunk 2 above)
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:396]: for single link, yes, but I think this will break some cases for galaxy. Generally though, this will need to be changed to be a lookup per channel. (Can you please call that instead - or add that API if it's missing?)

Comment (caixunshiren) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:396]: sounds good. I will apply this change

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp [hunk 4]
```diff
@@ -376,7 +381,7 @@ static void add_worker_config_to_edm_builders(
         }
 
         // Get the maximum message size we'd like to use. Not the actual packet size
-        uint32_t expected_message_size_bytes = tensor_slicer.get_worker_slice_size_bytes(global_worker_idx);
+        uint32_t expected_message_size_bytes = clockwise_edm_builders.at(link).get_eth_buffer_size_bytes();//tensor_slicer.get_worker_slice_size_bytes(global_worker_idx);
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:384]: Except for this line, this PR looks good. The small packet optimization should be kept in place.

**Comment by caixunshiren** [comment]
post commit: https://github.com/tenstorrent/tt-metal/actions/runs/10742171134
t3k frequent: https://github.com/tenstorrent/tt-metal/actions/runs/10742175010
t3k nightly: https://github.com/tenstorrent/tt-metal/actions/runs/10742179409

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp [hunk 5]
```diff
@@ -374,9 +379,6 @@ static void add_worker_config_to_edm_builders(
                 device->worker_core_from_logical_core(worker_cores.at(w)).y));
         }
 
-        // Get the maximum message size we'd like to use. Not the actual packet size
-        uint32_t expected_message_size_bytes = tensor_slicer.get_worker_slice_size_bytes(global_worker_idx);
```
[the remark this answers is not in the document]
Comment (caixunshiren) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:379]: updated it to default to num buffers = 1 and enable shrinking when num buffers = 1

**Review by SeanNijjar — APPROVED** [review]
